### PR TITLE
[MacOS][Pluggable Device] Fixes for macos pluggable device.

### PR DESCRIPTION
**Pull request by kulinseth**
1. Disable the Eigen max align check as we use custom device memory
   allocator in Metal plugin
2. Disable list kernels on Pluggable devices as Device memory on MacOS
   pluggable device doesn't allow us to index at a offset from the base
   MTLBuufer location.
3. Disable XLA compilation path as its not supported in pluggable
   architecture
cc. @penpornk , @learning-to-play , @nitins17

### CODE REVIEW

File: tensorflow/core/framework/tensor.h [hunk 1]
```diff
@@ -35,6 +35,12 @@ limitations under the License.
 #include "tensorflow/core/platform/mem.h"
 #include "tensorflow/core/platform/types.h"
 
+#if !defined(PLUGGABLE_DEVICE_SUPPORTED) &&                                \
```
Comment (penpornk) [inline tensorflow/core/framework/tensor.h:38]: Do you mean
```suggestion
#if !defined(PLUGGABLE_DEVICE_SUPPORTED_MACOS) &&                                \
```
?

File: tensorflow/core/framework/tensor.h [hunk 2]
```diff
@@ -35,6 +35,12 @@ limitations under the License.
...
+    (__x86_64__ || __i386__ || defined(__APPLE__)) && !defined(ANDROID) && \
+    !defined(__ANDROID__) && !TARGET_OS_IOS
```
Comment (penpornk) [inline tensorflow/core/framework/tensor.h:40]: These conditions are also for non-MacOS (any Linux on x86_64 and i386 will also satisfy the condition).
https://github.com/tensorflow/tensorflow/blob/5f7dcec3fabfa7b119790f87db20775088f0e2b7/tensorflow/core/kernels/list_kernels.h#L42-L49

To make it MacOS-specific, I think we only need
```suggestion
    defined(__APPLE__)) && !TARGET_OS_IOS
```

If you think this should apply to all PluggableDevice, then maybe we should rename it to something more generic.

**Review by penpornk — COMMENTED** [review]
Thank you for the PR!

@Jianhui-Li @jzhoulon @yiqianglee @PatriceVignola Would you like your plug-ins to disable Eigen max align check as well?

File: tensorflow/core/framework/tensor.h (hunk 1 above)
Comment (kulinseth) [reply tensorflow/core/framework/tensor.h]: Yes, as discussed, I have created a separate Macro for this.

File: tensorflow/core/framework/tensor.h (hunk 2 above)
Comment (kulinseth) [reply tensorflow/core/framework/tensor.h]: Done.

**Comment by Jianhui-Li** [comment]
> Thank you for the PR!
> 
> @Jianhui-Li @jzhoulon @yiqianglee @PatriceVignola Would you like your plug-ins to disable Eigen max align check as well?

I don't think we need this check for Intel Plugin. It seems to me an implementation issue specific to Metal plugin.  Even the device allocator doesnt' support alignment, the plugin could enforce alignment on top of it.

File: tensorflow/core/framework/tensor.h [hunk 3]
```diff
@@ -35,6 +35,12 @@ limitations under the License.
 #include "tensorflow/core/platform/mem.h"
 #include "tensorflow/core/platform/types.h"
 
+#if !defined(PLUGGABLE_DEVICE_SUPPORTED_MACOS) &&  \
+    defined(__APPLE__) && !defined(ANDROID) &&     \
+    !defined(__ANDROID__) && !TARGET_OS_IOS
```
Comment (penpornk) [inline tensorflow/core/framework/tensor.h:40]: Nit: `!defined(ANDROID)` appeared twice.
```suggestion
    defined(__APPLE__) && !defined(ANDROID) && !TARGET_OS_IOS
```

**Review by penpornk — CHANGES_REQUESTED** [review]
@Jianhui-Li Thank you for the quick reply!

File: tensorflow/core/framework/tensor.h (hunk 3 above)
Comment (penpornk) [reply tensorflow/core/framework/tensor.h]: @kulinseth  Sorry I misread ANDROID and __ANDROID__. I thought they were the same (just __ANDROID__ twice). I'll add this back internally!

**Review by penpornk — APPROVED** [review]
Thank you again!

**Pull request merged**